app/helpers/reactjs_helper.rb

```diff
@@ -11,58 +10,60 @@ def react_component(name, props = {})
     content_tag('foreman-react-component', '', :name => name, :data => { props: props })
   end
 
-  def webpacked_plugins_with_global_css
-    global_css_tags(global_plugins_list).join.html_safe
+  def js_tags_for(requested_plugins)
+    requested_plugins.map do |plugin|
+      javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{plugin.to_s.tr('-', '_')}','#{plugin.to_s.tr('-', '_')}','./index');".html_safe)
+    end
   end
 
   def webpacked_plugins_js_for(*plugin_names)
     js_tags_for(select_requested_plugins(plugin_names)).join.html_safe
   end
 
-  def webpacked_plugins_with_global_js
-    global_js_tags(global_plugins_list).join.html_safe
+  def other_webpack_plugin(plugin_name, file)
+    javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{plugin_name.to_s.tr('-', '_')}','#{plugin_name.to_s.tr('-', '_')}','./#{file}_index');".html_safe)
   end
 
-  def webpacked_plugins_css_for(*plugin_names)
-    css_tags_for(select_requested_plugins(plugin_names)).join.html_safe
+  def global_js_tags(requested_plugins)
+    requested_plugins.map do |plugin|
+      plugin[:files].map do |file|
+        javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{plugin[:id].to_s.tr('-', '_')}','#{plugin[:id].to_s.tr('-', '_')}','./#{file}_index');".html_safe)
+      end
+    end
   end
 
-  def select_requested_plugins(plugin_names)
-    available_plugins = Foreman::Plugin.with_webpack.map(&:id)
-    missing_plugins = plugin_names - available_plugins
-    if missing_plugins.any?
-      logger.error { "Failed to include webpack assets for plugins: #{missing_plugins}" }
-      raise ::Foreman::Exception.new("Failed to include webpack assets for plugins: #{missing_plugins}") if Rails.env.development?
-    end
-    plugin_names & available_plugins
+  def webpacked_plugins_with_global_js
+    global_js_tags(global_plugins_list).join.html_safe
   end
 
-  def js_tags_for(requested_plugins)
-    requested_plugins.map do |plugin|
-      javascript_include_tag(*webpack_asset_paths(plugin.to_s, :extension => 'js'))
-    end
+  def webpacked_plugins_css_for(*plugin_names)
+    Foreman::Deprecation.deprecation_warning('3.9', '`webpacked_plugins_css_for` is deprecated, plugin css is already loaded.')
   end
 
-  def global_js_tags(requested_plugins)
-    requested_plugins.map do |plugin|
-      plugin[:files].map do |file|
-        javascript_include_tag(*webpack_asset_paths("#{plugin[:id]}:#{file}", :extension => 'js'))
-      end
+  def get_webpack_foreman_vendor_js
+    if ENV['RAILS_ENV'] == 'production'
+      javascript_include_tag("/webpack/#{File.basename(Dir.glob('public/webpack/foreman-vendor*production*js')[0])}")
```

Comment:
Please avoid globbing for these methods. If there's some left over file it may end up using the wrong file. It also creates additional IO for every request at runtime. Previously we had some JSON file which had the mapping. Isn't that something that can still be done?

Reply:
Previously we had to create the JSON manifest files to make rails helpers to work correctly (it's a hack built upon the `webpack-rails` gem). Now that we don't use the gem and don't reference the files from Ruby code, I don't see a good reason to keep that special configuration in webpack to generate the manifest file. 
This concept inspired heavily b jsbundling-rails gem, which is the default option for rails7. The idea behind it is to reduce knowledge of the JS building process and serve the files as pure static files. 

In our specific case, there always should be only one vendor file in the folder, but we don't know the hash ID for it. Would it be better to assert single file here, and prevent the app from loading in case more than one file exists?

Reply:
Globbing can be unpredictable. sprockets also has a random component in its manifest filename and they at least have a warning if multiple files are found:

https://github.com/rails/sprockets/blob/5b040f3bec503ded8a98c0c911c7a77dd1f5bf1b/lib/sprockets/manifest_utils.rb#L40-L43

And you still haven't addressed my concern that this isn't stored in the application program, but loaded every time. I don't insist on a manifest, but your current approach causes additional IO and can make upgrades less predictable (because there can be a point where the new files are deployed, but the application not yet restarted).

I'm not sure what the best place is, but an application initializer would be my first thought.

Reply:
> And you still haven't addressed my concern that this isn't stored in the application program, but loaded every time. I don't insist on a manifest, but your current approach causes additional IO and can make upgrades less predictable (because there can be a point where the new files are deployed, but the application not yet restarted).

What about having the filename cached once?
Something along the lines of:
``` ruby  
  def self.manifest_file
    logger.warn("too many foreman-vendor files") if Dir.glob('public/webpack/foreman-vendor*production*js').length > 1
     @manifest_file ||= "/webpack/#{File.basename(Dir.glob('public/webpack/foreman-vendor*production*js')[0])}"
  end

  def get_webpack_foreman_vendor_js
      javascript_include_tag(self.manifest_file)
  end
```

This would make the upgrade predictable (until the next restart) and avoid extra IO.

Reply:
I would be worried about multiple workers. When Puma forks of a new worker, would it determine the manifest file again? In production mode I would expect it to be part of an initializer so it happens before forking.

And a small nit that I would use:
```ruby
def self.manifest_file
  @manifest_file ||= begin
    files = Dir.glob('public/webpack/foreman-vendor*production*js')
    logger.warn("too many foreman-vendor files") if files.length > 1
    "/webpack/#{File.basename(files.first)}"
  end
end
```

Because your logger statement would still glob on every request.

Reply:
@MariaAga  The whole idea is to get rid of the manifest, but if we decide to leave the manifest we shall use it of course.